src/cpp/tensorpack.cpp

```diff
@@ -30,152 +30,153 @@
             break;
-        case TensorType::int32 :
-            ptr = new Tensor<int32_t>(name, data, dims,
-                                      type, mem_layout);
+        case TensorType::int32:
+            ptr = new Tensor<int32_t>(name, data, dims, type, mem_layout);
             break;
-        case TensorType::int16 :
-            ptr = new Tensor<int16_t>(name, data, dims,
-                                     type, mem_layout);
+        case TensorType::int16:
+            ptr = new Tensor<int16_t>(name, data, dims, type, mem_layout);
             break;
-        case TensorType::int8 :
-            ptr = new Tensor<int8_t>(name, data, dims,
-                                     type, mem_layout);
+        case TensorType::int8:
+            ptr = new Tensor<int8_t>(name, data, dims, type, mem_layout);
             break;
-        case TensorType::uint16 :
-            ptr = new Tensor<uint16_t>(name, data, dims,
-                                       type, mem_layout);
+        case TensorType::uint16:
+            ptr = new Tensor<uint16_t>(name, data, dims, type, mem_layout);
             break;
-        case TensorType::uint8 :
-             ptr = new Tensor<uint8_t>(name, data, dims,
-                                       type, mem_layout);
+        case TensorType::uint8:
+            ptr = new Tensor<uint8_t>(name, data, dims, type, mem_layout);
              break;
         default :
-	  return; // FINDME: Handle this case better
+            throw std::runtime_error("Unknown tensor type");
     }
-    this->add_tensor(ptr);
-    return;
+
+    // Add it
+    add_tensor(ptr);
 }
 
+// Method to add a tensor object that has already been created on the heap.
+// DO NOT add tensors allocated on the stack that may be deleted outside of
+// the tensor pack. This function will cast the TensorBase to the correct
+// Tensor<T> type.
 void TensorPack::add_tensor(TensorBase* tensor)
 {
     std::string name = tensor->name();
 
-    if(name.size()==0)
-        throw std::runtime_error("The tensor name must "\
-                                 "be greater than 0.");
+    if (name.size() == 0)
+        throw std::runtime_error("The tensor name must be nonempty.");
 
-    this->_tensorbase_inventory[name] = tensor;
-    this->_all_tensors.push_front(tensor);
-    return;
+    _tensorbase_inventory[name] = tensor;
+    _all_tensors.push_front(tensor);
 }
 
+// Return a TensorBase pointer based on name.
 TensorBase* TensorPack::get_tensor(const std::string& name)
 {
-    TensorBase* ptr = this->_tensorbase_inventory.at(name);
-    return ptr;
+    return _tensorbase_inventory.at(name);
 }
 
+// Retrieve a pointer to the tensor data memory space
 void* TensorPack::get_tensor_data(const std::string& name)
 {
-    TensorBase* ptr = this->_tensorbase_inventory.at(name);
-    return ptr->data();
+    TensorBase* ptr = _tensorbase_inventory.at(name);
+    return (ptr == NULL ? ptr : ptr->data());
```

Comment:
I don't think this is the correct behavior.  Returning ``NULL`` implies that the data is ``NULL`` (since the function name is get_tensor_data()).  If we are worried that the value in the map is ``NULL``, then I think we should throw an error instead of return ``NULL``.